R/vector.R

```diff
@@ -0,0 +1,213 @@
+#' @return A modified vector upcasted to a list.
+#' @export
+#' @examples
+#' modify(c(1, b = 2, 3), 4, b = "foo")
+#'
+#' x <- list(a = 1, b = 2)
+#' y <- list(b = 3, c = 4)
+#' modify(x, .elts = y)
+modify <- function(.x, ..., .elts = list()) {
+  out <- as.list(.x)
+  args <- c(list(...), .elts)
+
+  args_nms <- names(args)
+  exists <- has_names(args) & args_nms %in% names(out)
+
+  for (nm in args_nms[exists]) {
+    out[[nm]] <- args[[nm]]
+  }
+
+  c(out, args[!exists])
+}
+
+
+#' Helper to create vectors with matching length.
+#'
+#' These functions take the idea of \code{\link{seq_along}} and
+#' generalise it to creating lists (\code{list_along}) and repeating
+#' values (\code{rep_along}). The dots are forwarded to
+#' \code{\link[base]{structure}()} to make it easy to add attributes.
+#'
+#' @inheritParams with_attributes
+#' @param .x A vector.
+#' @param .y Values to repeat.
+#' @examples
+#' x <- 0:5
+#' rep_along(x, 1:2)
+#' rep_along(x, 1)
+#' list_along(x)
+#'
+#' # You can also add attributes with additional arguments:
+#' rep_along(x, 1, class = "my_class")
+#' double_along(x, class = "my_class")
+#' @name along
+#' @seealso new-vectors
+NULL
+
+#' @export
+#' @rdname along
+logical_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("logical", length(.x)), ..., .attrs = .attrs)
+}
+#' @export
+#' @rdname along
+integer_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("integer", length(.x)), ..., .attrs = .attrs)
+}
+#' @export
+#' @rdname along
+double_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("double", length(.x)), ..., .attrs = .attrs)
+}
+#' @export
+#' @rdname along
+character_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("character", length(.x)), ..., .attrs = .attrs)
+}
+#' @export
+#' @rdname along
+list_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("list", length(.x)), ..., .attrs = .attrs)
+}
+#' @export
+#' @rdname along
+complex_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("complex", length(.x)), ..., .attrs = .attrs)
+}
+#' @export
+#' @rdname along
+raw_along <- function(.x, ..., .attrs = list()) {
+  with_attributes(vector("raw", length(.x)), ..., .attrs = .attrs)
+}
+
+#' @export
+#' @rdname along
+rep_along <- function(.x, .y, ..., .attrs = list()) {
+  with_attributes(rep(.y, length.out = length(.x)), ..., .attrs = .attrs)
+}
+
+
+#' Create new vectors.
+#'
+#' These functions construct vectors of given length, with attributes
+#' specified via dots.
+#'
+#' @inheritParams with_attributes
+#' @param .n The vector length.
+#' @examples
+#' list_until(10)
+#'
+#' # Add attributes, including the S3 class:
+#' integer_until(0, index = 1)
+#' double_until(10, class = "my_class")
+#' @name new-vectors
+#' @seealso along
+NULL
+
+#' @export
+#' @rdname new-vectors
+logical_until <- function(.n = 0, ..., .attrs = list()) {
```

Comment:
`_new()` felt like it would take actual contents rather than a length. Maybe that's just me?